### PR TITLE
feat: Introducing ability to track & navigate to windows from all applications across all spaces

### LINKED ISSUE #6: Dock peek: no visual indication that offscreen widow preview(s) exist
When dock peeking, a certain number of the app's window previews are shown on the screen. However, there is no visual indication to the user that additional window previews exist off screen (i.e., the user needs to scroll/swipe left to see them).

### LINKED ISSUE #9: Dock previews not available for apps running in another Space
Many similar Dock preview apps run into this problem.  Apparently apps like DockView and DockMate use a Private API to get the window preview for running windows in other Spaces (not the current Space).

Could you look into this?  Having only some windows show their Preview takes away from the mission of a Dock preview app.

### LINKED ISSUE #130: Window switcher hotkey some not update properly
Changing the hotkey for the window switcher does not update, and does not always recognize custom hotkeys. I wanted to use ⌥ + ⇥ but had to do ⌥ + ⇥ + shift, at which point it switched to ⌥ + ⇥.

While I'm here, a couple notes. The UI for the window switcher is very nice! A couple things I'd like to see tweaked would be having the windows be in order of use, not alphabetical. Staying open when closing an app from in the switcher would be useful, and having multiple layers when too many windows are open would be nice as well. Loving the app so far, hoping to fully replace a couple different ones I use!

### LINKED ISSUE #196: Window Switcher Order: Recently Focused
I thought I could replace AltTab with this more modern app, but a blocker is the window switcher order. I want to quickly and blindly switch between the last two apps back and forth (i.e. code editor + browser) with just a single keyboard shortcut. Setting AltTabs "Window order" to "Recently Focused" allows me to do this. The current app is at the start of the list, and the first app selected is the second last focused app. Pressing the shortcut quickly just opens the last app, as I also have a slight apparition delay of 150ms enabled which also means the window switcher itself isn't even shown for this quick back and forth switching, but I can live without that.

### LINKED ISSUE #232: [BUG] Window Preview showing Safari Downloads as a window when download list is open
## Describe the bug
If I have the downloads list open and preview Safari on Dock, it shows the downloads list as a safari window.

## Screenshots

Screenshot 2024-07-25 at 11 37 00 PM

## Environment
 - macOS version: Sonoma 14.5
 - App version: 1.1.4

### PR DESCRIPTION
**Pull request by hasansultan92**
## Describe your changes
Now able to track windows in other spaces. DockDoor is now able present windows in the window switcher and dock peek by the most recently used window.

No need to enable any settings, simply use your applications and the application will keep track of what window is open, modified, and closed/quit. 

Private api usage: Minimal to no use of private api

@ejbills huge thanks for assisting!
## Related issue number(s) and link(s)

This closes #232, closes #9, closes #196, closes #130, closes #6

## Checklist before requesting a review
- [.] I have performed a self-review of my code
- [.] If this change affects core functionality, I have added a description highlighting the changes
- [.] I have followed conventional commit guidelines

## Core Functionality Changes

1. We are now tracking windows individually and listening to the OS firing events/notifications for the windows modifying, created, and closed. 

1. We are also now able to sort the windows DockDoor has discovered and present them in the order of recently used. This allows the user to enable the window switcher with their favorite keybind which in return behaves similar to the CMD+Tab window switcher native to Mac OS.

1. One can now switch on the fly without having the window switcher preview pop up to their most recently used window

### CODE REVIEW

**Comment by ejbills** [comment]
@ShlomoCode this PR is done and is being tested at the moment. We have achieved the minimized/hidden window previews and feel this is working pretty well with minimal private API usage. Feel free to give this a try, I am wanting to give you a heads up and ask for feedback on how we can manage merging this while not setting your draft PR back too far.

**Comment by ShlomoCode** [comment]
> We have achieved the minimized/hidden window previews and feel this is working pretty well with minimal private API usage

I did a quick try (I haven't gone through the code yet), and the preview seems to freeze for minimized windows, and also doesn't show anything for windows minimized before DockDoor is launched, probably the trick is to take a quick picture before minimizing a window?
For me at least, this makes the feature pretty useless, as the window as it *now* cannot be peeked.

**Comment by ShlomoCode** [comment]
Same with full screen windows. It does not show a updated image, nor detact windows that were already in full screen before DockDoor's launch.

**Comment by ejbills** [comment]
That can be worked in after we merge this PR, since your changes that make that possible mainly come from the changes you made to how screenshots are taken. We can work it into our new caching mechanism and it'll have support for those.

**Comment by ShlomoCode** [comment]
Does this mean we are giving up windows that were minimized/full screen before launch?

**Comment by ejbills** [comment]
Wouldn't that also be covered by your PR? It seems like it completely reworks how windows are discovered, so we can simply wire it into the changes in this PR. I can help with that.

**Comment by ShlomoCode** [comment]
Precisely in this part there was no big change...
In my PR we no longer need to adjust windows according to title, but we have a `NSRunningApplication` so I also get the windows using `kAXChildrenAttribute` call, I don't think it has an advantage over the old method, we just need to check which is faster - loop on all windows and find (OLD) or children AX call on the Application AX element (NEW)

In summary: we can support already minimized windows (after using the private API to capture the window image), but not windows that were already in full screen before the launch

**Comment by ejbills** [comment]
I'm more favorable to the usage of ScreenCaptureKit where possible, and then of course also using private APIs where needed. But, I'm open to discussion on if we want to use the private APIs purely for the windows that existed pre-launch. Is the lack of private API usage worth losing out on that functionality? 

It's possible that mixing and matching official and private APIs will be harder to maintain than just to maintain the private API, which would cover all of our needs. Only, it would have the risk of breaking in future macOS releases.

**Comment by ShlomoCode** [comment]
It's not only for windows that were open before launch, it's also for a recent image of minimized windows

**Comment by ShlomoCode** [comment]
@ejbills I sent you a Google Chat invitation, please take a look

**Comment by hasansultan92** [comment]
Hey @ShlomoCode I'm not fully versed with the minimized windows issue but I am replying in response to the full-screen windows detached that are not being shown in the window switcher. This is actually how Mac OS seems to have it. if you are watching a YouTube video on Safari and in full screen you can't actually navigate to the video if you do a cmd + tab + arrow key on safari, it doesnt show up there and hence not on DockDoor either.

With regards to applications launched before DockDoor was launched and their windows, well Alt-Tab also discovers them when the user navigates to them and interacts with them, atleast thats how I recall this. I removed this functionality- [see this commit](https://github.com/ejbills/DockDoor/pull/234/commits/26c6f990241636efe8bdc33d4a2814806f438f53#diff-2154ad1f206d7cfd6b5aed0be8d4b21c6fc6326713e614edff887abdc4e6beda). Originally my thought process was that as the user moves around their computer, DockDoor would use that notification to keep track of the windows in our cache. I realized that logic was conflicting and we don't need the space changed notification handler since we can keep building the window cache on the fly as the user enables the window-switcher and dock peek. We can use the other notifications to remove and update the windows though. 

I'll be doing some more testing since there have been more changes since the last time I made a commit

**Comment by ShlomoCode** [comment]
It seems that alt-tab indeed detects windows in another space/full screen only after switching to them, and this is very strange because the code explicitly moves the windows to the current space for their initial detection:
https://github.com/lwouis/alt-tab-macos/blob/master/src/logic/Applications.swift#L38-L41

**Comment by hasansultan92** [comment]
Right. What I have gathered from Alt-Tabs source code is a couple of things. 

1. That project has existed for a while and tracing through their issues and commits, it seems like they have done quite a bit over the years to deal with the OS and its quirks/unexpected behaviors
2. Because they have existed for a while, they have employed methods that we do not need anymore
3. Because they have existed for a while, they are also technically backward compatible with older Mac OS which we are not. We require SCKit which is a Mac OS 14.0+ requirement - correct me if I am wrong.
4. What you have linked to employs the use of private apis, despite their program code compiling, there is also the possibility that that section of the program code just has no functional effect anymore over the years. Apple might have removed the underlying function contents and the compiler is just compiling some function with no body on what to do - not sure and please dont quote me on this.

So while I would have to navigate around to find my windows that launched before Alt-Tab, there is a possibility that someone on potentially older hardware and an older version of the OS might have once used an Alt-Tab where they would bring all the windows together to the current space and then brought them back to their original space after discovery and that never got deprecated/removed from the active file.

**Comment by ShlomoCode** [comment]
> SCKit which is a Mac OS 14.0+ requirement

https://developer.apple.com/documentation/screencapturekit
>macOS 12.3+

:thinking:

**Comment by ShlomoCode** [comment]
Ok, leaving aside the problem of existing windows before opening, the main drawback I see in this PR is the inability to get a recent image of a minimized/full screen window. However, as @ejbills wrote we can add this later using the private screenshot API.

Other than that I think it's fine, @hasansultan92 please see the few reviews I left.

**Comment by hasansultan92** [comment]
So we cannot get information about windows that are in different spaces before DockDoor launches. The reality is we need the AXUIElement for that and that API only works in the space you are in. I had a POC using the private apis that Alt-Tab uses in one of my branches from a couple of weeks back I believe and it was the same issue. I overhauled almost everything only to find out I need to build an AXUIElement for everything and I dont have access to that until we literally are in the same space as that application.

And thanks on the reviews, lemme have a read.

**Comment by ShlomoCode** [comment]
Minimized windows that exist before launch should work even without a private API, it doesn't work at the moment because we use ScreenCaptureKit which can't get an image of minimized windows
We should probably integrate the alt-tab private screenshot API, but we can leave that to my PR since it's already done there

**Comment by hasansultan92** [comment]
@ShlomoCode sorry I have not been able to find the review, were they left in the PR?

**Comment by ShlomoCode** [comment]
@hasansultan92 They are after this comment: https://github.com/ejbills/DockDoor/pull/234#issuecomment-2254691765

**Comment by hasansultan92** [comment]
hmm are we experiencing a github bug? 🤔

image

**Review by ShlomoCode — CHANGES_REQUESTED** [review]

**Comment by ShlomoCode** [comment]
@hasansultan92 I think now it should be visible

File: DockDoor/Utilities/WindowUtil.swift [hunk 1]
```diff
@@ -492,85 +529,148 @@ final class WindowUtil {
...
     
-    func wait() async {
-        if value > 0 {
-            value -= 1
-        } else {
-            await withCheckedContinuation { continuation in
-                waiters.append(continuation)
-            }
+    static func findAllWindowsInDesktopCacheForApplication(for applicationName: String) -> [WindowInfo]? {
+        let bundleID = findBundleID(for: applicationName)
```
[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift:646]: hmm, I was not aware that applications that spawn windows are allowed to exist without bundle ids - I thought all applications needed a bundle id. Do you happen to know of applications that exist in our day-to-day lives that I can use to append/alter this logic that don't happen to have a bundle id?

[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift:646]: Yes it's pretty rare, I know about android studio emulator and scrcpy

[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift:646]: Now that you mention this, I recall reading some comments in Alt-Tab about this. Was it just the emulator or the whole application? I am going to download it and do some testing with it to see the behavior and what else we might need to enhance.

[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift:646]: only the emulator

[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift:646]: Interesting ok. Let me download Android Studio and let me compare how the native OS is reacting. Depending on what the OS shows, ill try my best to match the same outcome

[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift:646]: Note that at https://github.com/ejbills/DockDoor/pull/214 I changed the approach to be more reliable in general (we work with `NSRunningApplication`)

File: DockDoor/Views/Hover Window/WindowPreview.swift [hunk 2]
```diff
@@ -57,45 +57,31 @@ struct WindowPreview: View {
...
+                            if isMinimized || isHidden {
+                                fluidGradient().opacity(0.225)
+                                HStack {
+                                    Image(systemName: "eye.slash.fill")
+                                    Text(isMinimized ? "Minimized" : "Hidden")
+                                }
+                                .font(.caption)
+                                .italic()
+                                .foregroundStyle(.secondary)
+                            }
```
[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Views/Hover Window/WindowPreview.swift:75]: @hasansultan92 what about that?

File: DockDoor/Utilities/WindowUtil.swift
Comment (ejbills) [inline DockDoor/Utilities/WindowUtil.swift:1]: Some windows can and won't have a close button. I suggest we move this guard.

Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift:1]: Is this comment in the right section?

File: DockDoor/Views/Hover Window/WindowPreview.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Views/Hover Window/WindowPreview.swift:75]: So this is more on the UI side of things, I was not aware we were having this issue to begin with, and something @ejbills has been contributing to mostly. I don't have an issue with this UI look but I do believe we should add more opacity to make the minimized window preview almost opaque

[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Views/Hover Window/WindowPreview.swift:75]: I don't understand why a preview of a minimized window should be blurry. In my opinion the main use of DockDoor is for minimized windows, normal windows are usually already visible

[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Views/Hover Window/WindowPreview.swift:75]: hmm, my thought would be the opposite. If I have a minimized window, that's because I don't want others to see or myself that it exists and I most certainly would not want to see the preview either, defeats the purpose of minimizing a window, though I also see your point. 

Perhaps we can introduce a state of some sort that can allow what you want and perhaps what I want? If the state is flipped we can have your preview, and if the state is true, we can have the preview I mentioned. Of course this can be vice versa too.

File: .github/workflows/draft_release.yml
Comment (ejbills) [inline .github/workflows/draft_release.yml:1]: Not sure why this is in the PR, this was removed.

File: DockDoor/Views/Hover Window/WindowPreview.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Views/Hover Window/WindowPreview.swift:75]: hmm, I see your point @ShlomoCode. Thoughts on this @ejbills?

File: DockDoor/Utilities/WindowUtil.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift:646]: Apologies this took a little while to get done, attached is a screenshot where we are able to pick up on the emulator and the IDE window as separate windows. I suggest we move forward with the bundle identifier, I am unaware of applications that can be compiled without a bundle Identifier for the apple ecosystem.
Screenshot 2024-08-01 at 13 57 50

**Comment by hasansultan92** [comment]
@ShlomoCode @ejbills can you guys tell me of the reference PWA that was used for #126? I would like to bring back the check and not allow a window to show up in our cache if it has no title.

File: DockDoor/Utilities/WindowUtil.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift:646]: This is on Mac OS Dev Beta 15.1 with Android Studio Koala. Not my Android project or anything, just one of the default ones they had

File: DockDoor/Views/Hover Window/WindowPreview.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (ejbills) [reply DockDoor/Views/Hover Window/WindowPreview.swift]: > I don't like this overlay ... It seems quite clunky

Don't get how it was clunky, but the UI was not final. I pushed a changed version which I am happy with, and will be the final UI for minimized/hidden windows. Open to constructive criticism of course.

@ShlomoCode this PR will be merged tomorrow pending some testing.

**Comment by ShlomoCode** [comment]
@hasansultan92 I installed the site https://www.geektime.co.il as a PWA (Chrome's 3 points > save and share > install the page as an application)

File: DockDoor/Utilities/WindowUtil.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift]: This is not an Android Studio emulator window. This is an Android Studio window. The Android Studio emulator is a completely separate floating window, you dragged out a part of Android Studio, "Running Devices". Also check out scrcpy.

Other than that, I don't see why we need to base on bundleID and convert back and forth to the appropriate information, when we have a much juicier information object like NSRunningApplication.

[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift]: Maybe put my change in the current under mouse app detection method into this PR? This way we can work with NSRunningApplication all the way through.

[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift]: https://developer.android.com/studio/run/emulator-launch-separate-window#:~:text=To%20launch%20the%20emulator%20in,the%20Running%20Devices%20tool%20window.

As of 2024-07-01 that is the Android Studio Emulator window. I also recall in the past that it would spawn a different style of window but I believe they have changed things around.

[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift]: To get that information object ```NSRunningApplication``` we need to make another call to get that information. We have the right information we need to perform the store/save operation and hence we don't need to make potentially another system call for information we already have. Storing the information is likely cheaper than making a system call because the OS is most likely keeping the application in the Cache. 

@ejbills correct me if I am wrong too.

[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift]: @ShlomoCode I don't have an Android device to fidget with scrcpy. Any chance you can pull the latest changes and see if it's not recognizing the window?

[the remark this answers is not in the document]
Comment (hasansultan92) [reply DockDoor/Utilities/WindowUtil.swift]: @ShlomoCode this is the emulator window. I tried to just launch it on an older version of Android Studio, I believe a June 2023 Giraffe release and still the same thing.

Screenshot 2024-08-02 at 10 37 12

[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift]: > @ShlomoCode I don't have an Android device to fidget with scrcpy.

scrcpy can connect to android studio emulator

[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift]: I pulled the PR and checked and it does not recognize a scrcpy window, even when it is not set to float above everything.

[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift]: >However, some emulator features are only available when you run it in a separate window. To launch the emulator in a separate window, go to...

***

> To get that information object `NSRunningApplication` we need to make another call to get that information

Why?
```
guard let runningApp = NSRunningApplication.runningApplications(withBundleIdentifier: bundleIdentifier).first else {
    print("Current App is not running (\(bundleIdentifier))")
    return nil
}
```

[the remark this answers is not in the document]
Comment (ejbills) [reply DockDoor/Utilities/WindowUtil.swift]: Since the current prod version of the app doesn't support apps without a bundle ID, I think this can be made as an enhancement in the future. For now, I will be merging the PR to get some users this feature and get testing going.

**Comment by stevehoek** [comment]
Love the addition of this feature, however it does not seem to work (no preview shown) on full screen remote-desktop style apps such as Microsoft Remote Desktop, RealVNC Viewer, Parsec.

**Comment by hasansultan92** [comment]
Hello @stevehoek, Could you please confirm if this issue is specific to full-screen applications? If you open your browser and play a YouTube video in full-screen mode, the preview does not appear or is not updated?

**Comment by stevehoek** [comment]
@hasansultan92 full screen browser windows do show up correctly in the dock preview

**Comment by hasansultan92** [comment]
Here is a screenshot of my open windows without the Microsoft Remote Desktop application in full-screen mode.
Screenshot 2024-08-03 at 19 57 43

Here is a screenshot of my open windows including the Microsoft Remote Desktop application in full-screen mode.
Screenshot 2024-08-03 at 19 59 04

What is the expectation you had @stevehoek?

**Comment by stevehoek** [comment]
@hasansultan92 I don't see any dock preview at all, for Microsoft Remote Desktop "connection center" or any of the remote desktop sessions.  I don't full-screen the connection center itself, only the connected sessions.

image

I do run the beta version of Microsoft Remote Desktop if that makes any difference.

**Comment by hasansultan92** [comment]
Try going to another space and then going into the dock and ask for the preview. I had a case like this for Microsoft Edge. Can you try to see if it works at all on the stable Microsoft Remote Desktop for you at all?

Screenshot 2024-08-03 at 20 07 27

**Comment by stevehoek** [comment]
When using the AppStore version of Microsoft Remote Desktop, I do see a preview for the connection center main window, but still not for my full-screen connections.  When I full-screen the connection center, the preview still shows the non-fullscreen representation.

No combination of switching to other spaces/windows helps.

FWIW, I do run AltTab as well (and disable DockDoor's switcher)

**Comment by hasansultan92** [comment]
Try to do CMD + Tab, choose the Microsoft Remote Desktop Application, and before you release CMD, press one of the arrow keys. Do you see your full screen windows there? 

If I go to Safari - a native application - and play a youtube video, then ask for full screen on the video, technically Safari has spawned a new window for me which is also a new "space" so to speak. If I do the CMD + Tab - native command - windows switcher, with the instructions I gave earlier, I cannot see that Youtube video full screen in one of the spaces. This is native behavior to not show full screen contents of a window.

If you can see your windows with the command I have suggested above, then this would be considered a bug potentially, otherwise, this is expected behavior.

Alt-Tab might be capable of showing a preview of the full screen windows. DockDoor is not Alt-Tab, and DockDoor does not use the logic Alt-Tab does.

With regards to the Beta application, not sure whats happening there. Did you launch both the Beta and the App store version of the application while DockDoor was open? I can see why one would give a preview and not the other then.

File: DockDoor/Utilities/WindowUtil.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift]: Makes sense

**Comment by stevehoek** [comment]
For what it is worth, it is now working due to a restart after the macOS 14.6 update!

There is one type of app that doesn't show a preview for me: iOS/iPadOS apps that are installed on Apple Silicon a Mac.
